src/workers/layouts/nearestParentVertical.ts

```diff
@@ -0,0 +1,444 @@
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+      if (competitorBelowConnections > competitorAboveConnections) {
+        // competitor has more below than above
+        // node wins, shove competitor down
+        await shove({
+          direction: 1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      // competitor has equal above and below, or node has more above
+      // place node above competitor
+      nodeShoveRecords.set(competingNodeId, -1)
+    }
+
+    if (nodeBelowConnections > nodeAboveConnections) {
+      // node has more below
+      if (
+        competitorBelowConnections > competitorAboveConnections && competitorBelowConnections > nodeBelowConnections
+      ) {
+        // competitor has more below than above, and more below than node
+        // node wins, shove competitor down
+        await shove({
+          direction: 1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      if (competitorAboveConnections > competitorBelowConnections) {
+        // competitor has more above than below
+        // node wins, shove competitor up
+        await shove({
+          direction: -1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      // competitor has equal above and below, or node has more below
+      // place node below competitor
+      nodeShoveRecords.set(competingNodeId, 1)
+    }
+
+    return await placeNearUpstreamNode(competingNodeId, nodeStartX)
+  }
+
+  function getNodeParentDirectionCounts(nodeId: string): [number, number] {
+    const node = message.data.nodes.get(nodeId)
+    const nodeRow = layout.get(nodeId)
+
+    if (!node || nodeRow === undefined) {
+      console.warn('NearestParentLayout: Node was not found in either the data or layout', nodeId)
+      return [0, 0]
+    }
+
+    return node.parents?.reduce((counts, parent) => {
+      const parentRow = layout.get(parent.id)
+
+      if (parentRow === undefined) {
+        console.warn('NearestParentLayout - getNodeParentDirectionCounts: Parent node not found on layout data', nodeId)
+        return counts
+      }
+
+      if (parentRow < nodeRow) {
+        counts[0] += 1
+      }
+
+      if (parentRow > nodeRow) {
+        counts[1] += 1
+      }
+
+      return counts
+    }, [0, 0]) ?? [0, 0]
+  }
+
+  function getNodeEndX(nodeId: string): number {
+    const nodeStartX = horizontal.get(nodeId)
+    const nodeWidth = message.widths.get(nodeId)
+
+    if (nodeStartX === undefined || nodeWidth === undefined) {
+      console.warn('NearestParentLayout: Node was not found in the horizontal layout and/or widths', nodeId)
+      return 0
+    }
+
+    return nodeStartX + nodeWidth
+  }
+
+  function purgeNegativeLayoutPositions(): void {
+    const allRows = Array.from( layout.values() )
+    const lowestRow = Math.min(...allRows)
```

Comment:
One micro optimization we could make for this is to track the lowest row as we go like `lowestRow = Math.min(lowestRow, newRow)`. That way we don't have to convert a map of potentially 10k values into an array here.

Reply:
Curious what you think of the solution I applied for this. Basically assigning a node row to the layout uses a function now instead of simply `layout.set()` so that the lowest row check can be made.